src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala

```diff
@@ -0,0 +1,363 @@
+package org.broadinstitute.dsde.workbench.sam.dataAccess
+
+import org.broadinstitute.dsde.workbench.sam.db.SamParameterBinderFactory._
+import org.broadinstitute.dsde.workbench.sam.db.tables._
+import org.broadinstitute.dsde.workbench.sam.model.{AccessPolicy, FullyQualifiedResourceId, ResourceTypeName}
+import scalikejdbc.{DBSession, _}
+
+/**
+  * Resources can specify access policies that affect their descendants. Checking these inherited policies
+  * at query time is expensive as it requires a recursive query. Effective policies are an optimization
+  * that shifts this burden from read time to write time by calculating and storing all policies that are
+  * effective for all resources so that at query time the lookup is straight forward, less stress on the db and fast.
+  *
+  * An effective policy applies to one and only one resource, never to descendants. It contains roles and actions
+  * that are applicable only to the resource; it is not required to check if roles are descendant only or apply
+  * to other resource types.
+  *
+  * It is crucial that all resource and policy updates are in serializable transactions to avoid race conditions when
+  * concurrent modifications are made affecting the same resource structure.
+  */
+trait EffectivePolicyMutationStatements {
```

Comment:
I found it helpful while reading through this to try and describe in my own words what a bunch of the different functions do. Leaving them in the review in case you want to include any as comments, but feel free to disregard if you don't find them helpful!